File: app/src/main/java/com/codingbuffalo/aerialdream/data/protium/ValueTask.java
package com.codingbuffalo.aerialdream.data.protium;

public abstract class ValueTask<T> extends Task<T> {
    @Override
    public final T call() {
        try {
            T data = onExecute();

            if (!Thread.currentThread().isInterrupted()) {
                onComplete(data);
                return data;
            }
        } catch (Exception e) {
            onError(e);
        }
        
        return null;
    }

    public abstract T onExecute() throws Exception;
    
    public void onComplete(T data) {
    }
}

File: app/src/main/java/com/codingbuffalo/aerialdream/data/protium/ValueTask.kt
package com.codingbuffalo.aerialdream.data.protium

abstract class ValueTask<T> : Task<T>() {
    override fun call(): T? {
        try {
            val data = onExecute()
            if (!Thread.currentThread().isInterrupted) {
                onComplete(data)
                return data
            }
        } catch (e: Exception) {
            onError(e)
        }
        return null
    }

    @Throws(Exception::class)
    abstract fun onExecute(): T
    open fun onComplete(data: T) {}
}